### PR TITLE
Support fp16 output from ImageInputOp

### PR DESCRIPTION
**Pull request by slayton58**
Refactors out option -> datatype cast
Allows native fp16 output from GPU transformation path
Expanded CastOp to allow fp16 in/out on GPU

### CODE REVIEW

File: caffe2/image/transform_gpu.cu [hunk 1]
```diff
@@ -62,4 +61,19 @@ bool TransformOnGPU<uint8_t, float, CUDAContext>(Tensor<CUDAContext>& X, Tensor<
   return true;
 }
 
+template <>
+bool TransformOnGPU<uint8_t, float16, CUDAContext>(Tensor<CUDAContext>& X, Tensor<CUDAContext> *Y, float std, float mean, CUDAContext *context)
```
Comment (asaadaldien) [inline caffe2/image/transform_gpu.cu]: The code is the same as above. What about moving the implementation to the templated version (line:45) ?

**Review by asaadaldien — CHANGES_REQUESTED** [review]
Looks good thanks @slayton58.
Just one comment :)

File: caffe2/image/transform_gpu.cu (hunk 1 above)
Comment (slayton58) [reply caffe2/image/transform_gpu.cu]: I swear there was a reason I didn't do that before, but I can't remember, and I fixed it this time, so it obviously wasn't too serious :)

Updated.

File: caffe2/image/image_input_op.h [hunk 2]
```diff
@@ -60,6 +61,7 @@ class ImageInputOp final
   bool is_test_;
   bool use_caffe_datum_;
   bool gpu_transform_;
+  TensorProto_DataType output_type_;
```
Comment (asaadaldien) [inline caffe2/image/image_input_op.h]: Move this line bellow line:68 (breaks initialization order rule).

**Review by asaadaldien — CHANGES_REQUESTED** [review]

File: caffe2/image/image_input_op.h (hunk 2 above)
Comment (slayton58) [reply caffe2/image/image_input_op.h]: Done.

**Pull request closed without merging**